synapse/rest/client/room.py

```diff
@@ -1070,6 +1070,32 @@ def register_txn_path(
         )
 
 
+class TimestampLookupRestServlet(RestServlet):
+    PATTERNS = client_patterns("/rooms/(?P<room_id>[^/]*)/timestamp_to_event$")
+
+    def __init__(self, hs: "HomeServer"):
+        super().__init__()
+        self._auth = hs.get_auth()
+        self._store = hs.get_datastore()
+
+    async def on_GET(
+        self, request: SynapseRequest, room_id: str
+    ) -> Tuple[int, JsonDict]:
+        requester = await self._auth.get_user_by_req(request)
+        await self._auth.check_user_in_room(room_id, requester.user.to_string())
+
+        timestamp = parse_integer(request, "ts")
+        direction = parse_string(request, "dir", default="f", allowed_values=["f", "b"])
+
+        event_id = await self._store.get_event_for_timestamp(
+            room_id, timestamp, direction
+        )
+
+        return 200, {
+            "event_id": event_id,
+        }
```

Comment:
One problem I'm seeing with going with a `/timestamp_to_event` endpoint is that there is no way to actually paginate with `/messages` after getting the `event_id`.

We probably want to consider returning a pagination stream token to use with the `from`/`to` on `/messages`

Reply:
Or just not worry about it, we can make a separate request to get one:

```go
// Get a pagination token for a single event
contextRes := c.MustDoFunc(t, "GET", []string{"_matrix", "client", "r0", "rooms", roomID, "context", someEventId}, client.WithContentType("application/json"), client.WithQueries(url.Values{
	"limit": []string{"0"},
}))
```